_posts/2020-10-27-a-closer-look-at-the-interpreter.md

```diff
@@ -0,0 +1,202 @@
+
+The [`beam_makeops`] script takes this definition and generates code for
+the parts that are very cumbersome to write by hand, such as jumping to the
+next instruction. It also hides argument handling which allows us to reduce
+the code size by packing small arguments together behind the scenes, which
+we've explored in an [earlier post] on the subject.
+
+For performance reasons it also generates different variants based on the
+argument types outlined in [`ops.tab`] to reduce the amount of work we need to
+do at runtime.
+
+Let's have a look at the generated code for this instruction, specifically the
+variant for `X` registers:
+
+```c++
+/* (This has been modified slightly for readability) */
+case is_nonempty_list_fx:
+{
+    Eterm arg_word, term;
+
+    /* Read the argument word from the instruction
+     * stream. */
+    arg_word = I[1];
+
+    /* Unpack the offset of our source register (upper
+     * 32 bits) and then read its contents.
+     *
+     * Note that we address the X registers directly;
+     * had this instruction not been specialized, we
+     * would first need to determine whether the
+     * argument was an X or a Y register. */
+    term = x_registers[arg_word >> 32];
+
+    /* is_not_list(term) */
+    if (term & (_TAG_PRIMARY_MASK - TAG_PRIMARY_LIST)) {
+
+        /* Unpack our fail label (lower 32 bits) and add
+         * it to our current instruction pointer. This
+         * is an offset and may be negative for backward
+         * jumps. */
+        I += (Sint32)arg_word;
+
+        /* Jump to the instruction at the fail label using
+         * the "labels as values" GCC extension. */
+        goto *I;
+    }
+
+    /* Skip the current label address and argument word,
+     * then jump to the next instruction. */
+    I += 2;
+    goto *I;
+}
+```
+
+There's a bit more to it, and those who would like to know more can read the
+[documentation for the `beam_makeops` script], but the above covers the gist of
+it.
+
+While the above is quite efficient, there's significant overhead from dispatch
+and argument handling. Here's a slightly altered assembly listing of the
+above:
+
+<pre class="highlight">
+    <em>; Load the argument word and extract the offset of our
+    ; source register, which lives in the upper 32 bits of
+    ; said word.</em>
+    mov    rdx, [rbx + 8]
+    mov    rcx, rdx
+    shr    rcx, 32
+
+    <em>; Our source register is at address [r15 + rcx]. Check
+    ; whether it contains a list. </em>
+    <b>test   byte [r15 + rcx], (_TAG_PRIMARY_MASK - TAG_PRIMARY_LIST)
+    jne    jump_to_fail_label</b>
+
+    <em>; Jump to the instruction directly following this one.</em>
+    add    rbx, 16
+    jmp    [rbx]
+
+jump_to_fail_label:
+    <em>; Extract the offset of our fail label from the
+    ; argument word. It lives in the lower 32 bits so
+    ; we'll sign-extend it to the full 64 bits.</em>
+    movsxd rdx, edx
+
+    <em>; Add our offset to the instruction pointer and then
```

Comment:
I think it would be nice to make the wording similar to the comment in C code in the corresponding place.